kicad: add option to override footprint, symbol and 3d model locations

matthuszagh opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This PR makes it easy to supply custom library locations, which is useful when working on the libraries or using a git version.

I'm not really sure what the standard nixpkgs practice is here. I like this solution because it makes it really easy to work on the various libraries or use a git version. As I see it, the two main alternatives to this are:

1. Do it in an overlay.
2. Do it manually with the "Manage Symbol Libraries..." and "Manage Footprint Libraries...", etc. GUI menu options.

I don't like option 1 because correctly doing this with an overlay does not seem trivial, although maybe someone better at Nix will show me why it's easy. I'm also not a fan of option 2 because the GUI options don't allow you to specify a path that Kicad looks in for the libraries. Rather, you have to specify each library/footprint file individually which is a lot of unnecessary work when doing a lot of work on the kicad libraries. Option 2 also doesn't help when you want to use the git master version of a library.

What are people's thoughts on this?

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

matthuszagh commented:
Another comment: the whole kicad expression is really tough to override. For instance, changing `src` does nothing since it's set to `base`. But `base` is in a let expression so can't be changed.

matthuszagh commented:
I'm closing this as I'm not convinced my PR is a good idea and my comment without an associated patch isn't super useful. I do think there are problems with this package expression that should be addressed, but I'll hold off until I have some actual changes that others can review. Of course if you disagree feel free to reopen.

evils commented:
huh, not sure why i wasn't added as a reviewer

i did not consider ease of use of mixed components
i suspect it's rare enough that the "do it manually" option is adequate
i think my approach would be to change versions.nix if i wanted to change the library used
#82634 moves libraries and base out of the let block, which may allow such a thing from nix

i overhauled most of the kicad package a while ago but didn't receive much feedback
the previous approach was to create links which left the env var way of overwriting those paths open
this unfortunately tied the compiled source to specific libraries (meaning i had to recompile every time i wanted to try another wrapping related change)
this remains the approach for the i18n because that can't be supplied as an env var

@matthuszagh if you end up reviewing the package, please tag me, or try me on IRC

matthuszagh commented:
@evils I'm reopening this, and will fix merge conflicts. I'd like to get your input. I know you mentioned that you would override versions.nix. However, because versions is imported in a let block, I don't see a way to overlay this. Maybe I'm missing something easy though? Additionally, and more importantly in my mind, this still suffers from the library files not being writable.

Adding path arguments, as this PR does, provides a simple way to get either behavior (yours by default). Do you see any significant downside to it?

evils commented:
> you mentioned that you would override versions.nix. However, because versions is imported in a let block, I don't see a way to overlay this.

i meant my approach if i wanted to mix and match libraries would be to edit the `versions.nix` file in my nixpkgs checkout (i hope there is a nix native way to change either the content or which file is used, but haven't considered that)  
or if i had a local library i wanted to use, add it to kicad via the GUI  

> this still suffers from the library files not being writable

as long as the library files are distributed by nix, that's always going to be the case?
i don't think the library files are supposed to be writable, my approach if i want to change a something in them is to make add a modified version as a project specific library, which accompanies the project and is writable

> Do you see any significant downside to it?

if it fulfills the needs of a user that is not met by the current package, it's probably worth a bit of 'clutter'

matthuszagh commented:
Thanks for the feedback @evils!

> as long as the library files are distributed by nix, that's always going to be the case?

Agreed, and that makes sense. The nice thing about this PR is kicad doesn't need the actual files, it just needs the path string so doing this doesn't create any purity issues. For instance, I'm using this with flakes and it works fine.

> if it fulfills the needs of a user that is not met by the current package, it's probably worth a bit of 'clutter'

Yes, that it does.

matthuszagh commented:
Rebased for [this PR](https://github.com/NixOS/nixpkgs/pull/100066).

matthuszagh commented:
@doronbehar if you don't mind, it would be great to get your feedback on this PR too.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -197,13 +200,16 @@ stdenv.mkDerivation rec {
     "--prefix XDG_DATA_DIRS : ${cups}/share"
     "--prefix GIO_EXTRA_MODULES : ${gnome3.dconf}/lib/gio/modules"
 
-    "--set KISYSMOD ${footprints}/share/kicad/modules"
```
You could get away with a rebuild if you'd use the same order of `makeWrapperArgs` for the default build configuration.

Also, it would be better to set the value up above under `let`, and use it here.

doronbehar reviewed (changes requested):
I take back my approval.

matthuszagh commented:
@doronbehar changes done. Let me know.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -187,27 +190,34 @@ stdenv.mkDerivation rec {
+  makeWrapperArgs =
+    let
+      # Use user-provided symbol, footprint and 3D package locations if set.
+      symbols = if symbolsPath != null then symbolsPath else symbols;
```
You have a whole `let in` section up there were you can do what you want.

matthuszagh replied:
Actually my commit shouldn't work because `symbols` etc aren't defined here. I need `passthru.libraries.symbols`. But won't this prevent doing this in the upper let? That's why I placed this here.

doronbehar replied:
That's why `with` is an anti pattern. Get rid of it, and put in `passthru.libraries.symbols` the real value according to the override.

matthuszagh replied:
Ok I've made a change. I'm not confident I've achieved what you're asking for. Would you prefer that I change `libraries.nix` so that symbols can be defined directly in `passthru.libraries`?

matthuszagh replied:
I think changing `libraries.nix` is the only way to be able to define `passthru.libraries.symbols=...` to incorporate the override, but maybe I'm missing something?

doronbehar replied:
> I think changing `libraries.nix` is the only way to be able to define `passthru.libraries.symbols=...`

It's not. _If_ this was what I'd have recommend to do, you'd have had to use something like:

```nix
passthru.libraries = (callPackages ./libraries.nix { }) // { symbols = ...; };

doronbehar commented:
@matthuszagh, this is all wrong, because `makeWrapper` supports:

https://github.com/NixOS/nixpkgs/blob/62e614262ad4b9df3d40693350f516794e95ecda/pkgs/build-support/setup-hooks/make-wrapper.sh#L18-L19

So I'd suggest to make this PR do nothing besides switching the `--set` arguments to `--set-default`. Please explain your use case in the commit message.

matthuszagh committed:
kicad: set default footprint, symbol library and package locations

Previously, these library locations were set absolutely. This
prevented overriding their locations with environment variables. Now,
setting the corresponding environment variable will override the
setting in the environment wrapper. For instance, I can set

KISYSMOD=/some/path/to/footprints

and this will be used as my footprint library instead of the default
footprint library in the nix store. This feature is particularly
useful for having kicad libraries which are writable.

matthuszagh commented:
@doronbehar change done.

doronbehar commented:
Thanks for your contribution.
